# Wine Matching

-------------------------------------------------------------------------------------------------------------------------------

## Bagian Pertama : Pembukaan
Wine Matching adalah program yang bertujuan untuk memeberikan list wine yang ingin dicari berdasarkan inputan yang dimasukkan. Program ini kami bangun dilatar belakangi oleh keingintahuan untuk mendalami Pemrosesan Bahasa Alami dan mendapatkan nilai proyek akhir dari matakuliah Pemrosesan Bahasa Alami.

### Oleh Kelompok Oenophile
1. 11S16027 Haris Situmeang
2. 11S16032 Achmad Nasution
3. 11S16034 Dohar Napitupulu
4. 11S16055 Samuel Napitupulu

### Rincian Notebook
Notebook ini berisi tentang bagaimana kelompok kami melakukan implementasi dari program Wine Matching. Notebook ini juga menjadi laporan utama dari proyek akhir mata kuliah Pemrosesan Bahasa Alami dengan dosen pengampu:
1. Samuel Situmeang, S.Ti., M.Sc
2. Yaya Setiadi, S.Si., M.T
dan teaching assistant yang selalu menemani saat praktikum berlangsung:
1. Dwide Sembiring, S.Kom
2. Romasi Sibarani, S.Kom
Notebook ini akan dibagi menjadi beberapa bagian:

#### 1. Bagian Pertama : Pembukaan
Pada bagian ini akan dilakukan penjelasan singkat dari program, perkenalan kelompok, dan juga rincian dari notebook ini.

#### 2. Bagian Kedua : Load Data
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami melakukan load data Wine yang akan kami oleh untuk menjadi data utama dari program yang akan kami bangun. Data Wine yang kami gunakan berasal dari Kaggle
https://www.kaggle.com/christopheiv/winemagdata130k
Data ini membicarakan tentang Review dan Description sebuah produk Wine yang berasal dari website Winemag
https://www.winemag.com/

#### 3. Bagian Ketiga : Data Cleaning
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami melakukan cleaning data Wine yang sudah ada. Bagian ini bertujuan untuk memastikan kejanggalan data tidak mempengaruhi hasil dari Wine Matching yang dilakukan.

#### 4. Bagian Keempat : Text Preprocessing
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami melakukan text preprocessing data Wine yang sudah ada. Bagian ini bertujuan untuk memastikan data bersih dari keambiguitasan kata.

#### 5. Bagian Kelima : Feature Extraction
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami melakukan feature extraction pada data Wine. Feature extraction diperlukan pada wine matching ini dikarnakan feature yang didapat akan dijadikan acuan dalam pencarian nantinya.

#### 6. Bagian Keenam : Calculating The Pairwise Similarity Scores
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami melakukan calculating the pairwise similarity scores pada data features yang sudah didapat dari bagian sebelumnya. Calculating the pairwise similarity scores ini berguna dalam mendapatkan nilai kemiripan dari data features.

#### 7. Bagian Ketujuh : Program
Pada bagian ini akan dijelaskan dan diimplementasikan bagaimana kelompok kami membangun program Wine Matching.

 ------------------------------------------------------------------------------------------------------------------------------

## Bagian Kedua : Load Data
Load data yang kelompok kami lekukan dibantu oleh pandas library, pandas adalah sebuah librari berlisensi BSD dan open source yang menyediakan struktur data dan analisis data yang mudah digunakan dan berkinerja tinggi untuk bahasa pemrograman Python. Data akan diload pada variabel yang dinamai sebagai "data". 

Cara load data yang kelompok kami lakukan adalah dengan menggunakan fungsi read_csv yang disedikaan oleh pandas library.

In [3]:
import pandas as pd 

data = pd.read_csv("D:\PBA\project\winemag-data-130k-v2.csv")
data.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [4]:
data.shape

(129971, 14)

Dapat kita lihat data yang diload memiliki 129971 row dan 14 column. Hal ini membuktikan bahwa data yang kami gunakan sudah memenuhi syarat dari proyek akhir matakuliah Pemrosesan Bahasa Alami yaitu diharuskanya menggunakan data yang berjumlah 100000 lebih.

 ------------------------------------------------------------------------------------------------------------------------------

## Bagian Ketiga : Data Cleaning
Seperti yang sudah dijelaskan pada rincian, bagian ini bertujuan untuk mengolah kejanggalan data agar tidak mempengaruhi hasil dari program Wine Matching yang ingin dibagun. Bagian ini dibagi menjadi beberapa tahapan yakni sebagai berikut:

### 1. Tahap Pertama:  Droping Duplicates Data
Tahapan ini bertujuan untuk menangani duplicates data yang ada. Duplicates data yang akan didrop adalah berdasarkan column title yang ada pada data.

Cara mencari data yang duplicate adalah dengan menggunakan fungsi duplicated yang sudah ada python.

In [5]:
data[data.duplicated(['title'], keep=False)].head()

Unnamed: 0 country                                        description  \
9            9  France  This has great depth of flavor with its fresh ...   
10          10      US  Soft, supple plum envelopes an oaky structure ...   
11          11  France  This is a dry wine, very spicy, with a tight, ...   
12          12      US  Slightly reduced, this wine offers a chalky, t...   
13          13   Italy  This is dominated by oak and oak-driven aromas...   

       designation  points  price           province          region_1  \
9      Les Natures      87   27.0             Alsace            Alsace   
10  Mountain Cuvée      87   19.0         California       Napa Valley   
11             NaN      87   30.0             Alsace            Alsace   
12             NaN      87   34.0         California  Alexander Valley   
13           Rosso      87    NaN  Sicily & Sardinia              Etna   

   region_2     taster_name taster_twitter_handle  \
9       NaN      Roger Voss            @vossroger   
10     Napa  Virginie Boone               @vboone   
11      NaN      Roger Voss            @vossroger   
12   Sonoma  Virginie Boone               @vboone   
13      NaN   Kerin O’Keefe          @kerinokeefe   

                                                title             variety  \
9   Jean-Baptiste Adam 2012 Les Natures Pinot Gris...          Pinot Gris   
10  Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   
11            Leon Beyer 2012 Gewurztraminer (Alsace)      Gewürztraminer   
12  Louis M. Martini 2012 Cabernet Sauvignon (Alex...  Cabernet Sauvignon   
13             Masseria Setteporte 2012 Rosso  (Etna)   Nerello Mascalese   

                 winery  
9    Jean-Baptiste Adam  
10   Kirkland Signature  
11           Leon Beyer  
12     Louis M. Martini  
13  Masseria Setteporte

Sedangkan cara untuk mendrop data yang duplicate itu sendiri adalah dengan menggunakan fungsi drop_duplicates yang disediakan python.

In [6]:
data.drop_duplicates(subset=['title'], keep=False, inplace=True)
data.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [7]:
data.shape

(108104, 14)

Dapat kita lihat, data sekarang hanya memiliki 108104 row. Hal ini dikarnakan duplicates data yang sudah didrop dari data yang ingin kita proses selanjutnya.

### 2. Tahap Kedua : Dealing with Null Values
Tahapan ini bertujuan untuk menangani data yang miliki null values pada columnnya.

Cara untuk mencari null values pada data adalah dengan cara yang umum yaitu melakukan tracing data sambil mengecek apakah data tersebut null values atau tidaknya dengan menggunakan fungsi isnull.

In [8]:
data.isnull().sum()

Unnamed: 0                   0
country                     55
description                  0
designation              31164
points                       0
price                     7556
province                    55
region_1                 17754
region_2                 65399
taster_name              23293
taster_twitter_handle    27368
title                        0
variety                      1
winery                       0
dtype: int64

Terlihat beberapa column hanya memiliki sedikit null values saja dan ada juga yang memiliki banyak nilai null values. Kami akan mendrop data null values yang ada pada column yang sedikit memiliki null values saja. Dan mendrop column yang memiliki banyak null values dikarnakan pastinya akan mengganggu hasil dari program Wine Mathcing nantinya.

Cara untuk mendrop null values adalah dengan menggunakan fungsi dropna yang disediakan oleh python.

In [9]:
data = data.dropna(subset=['country'])
data = data.dropna(subset=['province'])
data = data.dropna(subset=['price'])
data = data.dropna(subset=['variety'])

Sedangkan untuk cara mendrop column adalah dengan menggunakan fungsi drop yang sudah disediakan oleh python.

In [10]:
data = data.drop(['designation', 'region_1', 'region_2','taster_name', 'taster_twitter_handle' ], axis=1)
data.head()

Unnamed: 0   country                                        description  \
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   
5           5     Spain  Blackberry and raspberry aromas show a typical...   

   points  price        province  \
1      87   15.0           Douro   
2      87   14.0          Oregon   
3      87   13.0        Michigan   
4      87   65.0          Oregon   
5      87   15.0  Northern Spain   

                                               title             variety  \
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)      Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)          Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...            Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...          Pinot Noir   
5  Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...  Tempranillo-Merlot   

                winery  
1  Quinta dos Avidagos  
2            Rainstorm  
3           St. Julian  
4         Sweet Cheeks  
5               Tandem

In [11]:
data.isnull().sum()

Unnamed: 0     0
country        0
description    0
points         0
price          0
province       0
title          0
variety        0
winery         0
dtype: int64

In [12]:
data.shape

(100496, 9)

Dapat kita lihat, data berkurang kembali dikarnakan kita sudah melakukan drop data yang memiliki null values pada columnnya. Data yang sekarang sudah menjadi 9 column saja. Dapat disimpulkan juga, walaupun sudah melakukan data cleaning, data yang kami gunakan masih mencakupi syarat dari proyek akhir matakuliah Pemrosesan Bahasa Alami.

Dikarnakan 100496 sangat banyak dan membuat laptop crash, kelompok kami hanya akan menggunakan 10000 data saja sebagai sample.

In [13]:
data = data.sample(n=10000, random_state=42)
data.head()

Unnamed: 0 country                                        description  \
101350      101350   Italy  This blend of Sangiovese, Cabernet Sauvignon a...   
118867      118867      US  Fulfills one of the main tenets of California ...   
58193        58193      US  Buttered black cherries and olallieberries mix...   
2255          2255      US  Rich, savory and appealing now for its smooth ...   
105911      105911      US  There are just enough plush, woody tannins her...   

        points  price    province  \
101350      87   55.0     Tuscany   
118867      90   28.0  California   
58193       88   58.0  California   
2255        87   25.0  California   
105911      83   17.0    Virginia   

                                                    title             variety  \
101350  Fattoria la Castellina 2008 Aureo Red (Rosso d...           Red Blend   
118867  Sinor-LaVallee 2010 Les Galets Vineyards Syrah...               Syrah   
58193               Proulx 2013 Dimples Red (Paso Robles)           Red Blend   
2255              X 2007 Cabernet Sauvignon (Napa Valley)  Cabernet Sauvignon   
105911            Cooper Vineyards 2007 Norton (Virginia)              Norton   

                        winery  
101350  Fattoria la Castellina  
118867          Sinor-LaVallee  
58193                   Proulx  
2255                         X  
105911        Cooper Vineyards

In [14]:
data.shape

(10000, 9)

------------------------------------------------------------------------------------------------------------------------------- 

## Bagian Keempat : Text Preprocessing
Seperti yang sudah dijelaskan pada bagian rincian notebook, bagian ini bertujuan untuk memastikan data terhindar dari keambiguitasan kata agar tidak mempengaruhi hasil dari program Wine Matching yang ingin dibangun. Bagian ini dibagi menjadi beberapa tahapan sebagai berikut:

### 1. Tahap Pertama : Stemming
Tahapan stemming yang kami lakukan dibantu oleh SnowballStemmer yang berasal dari nltk library. Kelompok kami memilih menggunakan SnowballStemmer dikarnakan berdasakarnakan penjelasan https://kite.com/python/docs/nltk.SnowballStemmer
SnowballStemmer adalah optimasi dari PorterStemmer menjadikan kelompok kami berasumsi bahwa SnowballStemmer lebih unggul dan juga open source yang kami gunakan sebagai referensi menggunakan ini. Contoh keunggulan SnowballStemmer terdapat pada website ini https://www.nltk.org/howto/stem.html

In [15]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer('english')

### 2. Tahap Kedua : Tokenizing

Kelompok kami juga memilih menggunkaan RegexpTokenizer dikarnakan kami menghindari kata yang diawali dengan angka menjadi 1 token.

Cara kelompok kami melakukan tokenizing adalah dengan:
1. melakukan tokenizing dengan regex yang sudah ada
2. token yang dihasilkan dilowerkan menggunakan fungsi lower yang ada pada python
3. melakukan stemming pada setiap token yang dihasilkan

In [16]:
from nltk.tokenize import RegexpTokenizer
tokenizer = RegexpTokenizer(r'[a-zA-Z\']+')

def tokenize(text):
    return [stemmer.stem(word) for word in tokenizer.tokenize(text.lower())]

 ------------------------------------------------------------------------------------------------------------------------------

## Bagian Kelima : Feature Extraction
Kelompok kami menggunakan Term Frequency Inverse Document Frequency (TF-IDF) dalam melakukan feature extraction. Cara kerja utama dari TF-IDF itu sendiri adalah menurunkan bobot kata-kata yang lebih umum. Sehingga pemodelan mendapatkan hasil yang lebih optimal dari kata-kata yang digunakan pada Wine Matching yang ingin dibagun. Bagian ini dibagi menjadi beberapa tahapan sebagai berikut:

### 1. Tahap Pertama : Stopwords Removal
Dikarnakan data berisi text dengan bahasa inggris, kelompok kami akan menggunakan stopwords yang berbahasa inggris juga yang dibantu oleh ENGLISH_STOP_WORDS yang terdapat pada FeatureExtraction dalam sklearn library.

In [17]:
from sklearn.feature_extraction import text
stop_words = text.ENGLISH_STOP_WORDS

### 2. Tahap Kedua : TF-IDF
Kelompok kami melakukan TF-IDF dengan bantuan dari TfidfVectorizer yang terdapat pada FeatureExtraction dalam sklearn library. Kelompok kami juga akan menggunakan column description sebagai column yang menurut kami adalah column yang dapat mewakili seluruh clolumn yang ada.

Kelompok kami juga hanya akan mengambil 1000 feature saja untuk menjadi sample.

In [18]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf = TfidfVectorizer(stop_words=stop_words, tokenizer=tokenize, max_features=1000)
tf_matrix = tf.fit_transform(data['description'])

C:\Users\User\Anaconda3\lib\site-packages\sklearn\feature_extraction\text.py:300: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['abov', 'afterward', 'alon', 'alreadi', 'alway', 'ani', 'anoth', 'anyon', 'anyth', 'anywher', 'becam', 'becaus', 'becom', 'befor', 'besid', 'cri', 'describ', 'dure', 'els', 'elsewher', 'empti', 'everi', 'everyon', 'everyth', 'everywher', 'fifti', 'forti', 'henc', 'hereaft', 'herebi', 'howev', 'hundr', 'inde', 'mani', 'meanwhil', 'moreov', 'nobodi', 'noon', 'noth', 'nowher', 'onc', 'onli', 'otherwis', 'ourselv', 'perhap', 'pleas', 'sever', 'sinc', 'sincer', 'sixti', 'someon', 'someth', 'sometim', 'somewher', 'themselv', 'thenc', 'thereaft', 'therebi', 'therefor', 'togeth', 'twelv', 'twenti', 'veri', 'whatev', 'whenc', 'whenev', 'wherea', 'whereaft', 'wherebi', 'wherev', 'whi', 'yourselv'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


In [19]:
tf_matrix.shape

(10000, 1000)

 ------------------------------------------------------------------------------------------------------------------------------

## Bagian Keenam : Calculating The Pairwise Similarity Scores
Seperti yang sudah dijelaskan pada rincian, bagian ini akan menghitung pairwise similarity scores. Kelompok kami akan menggunakan Cosine Similarity dalam menghitung pairwise seimilarity scores. Kelompok kami mendapatkan bantuan untuk melakukan Cosine Similarity dari linear_kernel dari sklearn library.

In [20]:
from sklearn.metrics.pairwise import linear_kernel
cosine_sim = linear_kernel(tf_matrix, tf_matrix)

In [21]:
cosine = pd.DataFrame(cosine_sim)
cosine.head()

0         1         2         3         4         5         6     \
0  1.000000  0.030766  0.075663  0.000000  0.071534  0.057926  0.061480   
1  0.030766  1.000000  0.034789  0.024070  0.033599  0.035565  0.025355   
2  0.075663  0.034789  1.000000  0.026701  0.018430  0.028838  0.000000   
3  0.000000  0.024070  0.026701  1.000000  0.083349  0.044421  0.032140   
4  0.071534  0.033599  0.018430  0.083349  1.000000  0.027048  0.016099   

       7         8         9     ...      9990      9991      9992      9993  \
0  0.063950  0.172744  0.044490  ...  0.000000  0.180422  0.037886  0.184209   
1  0.160380  0.053754  0.022365  ...  0.100893  0.097161  0.095305  0.046278   
2  0.062853  0.025289  0.015658  ...  0.033890  0.171605  0.078498  0.000000   
3  0.070208  0.010716  0.037750  ...  0.009195  0.066644  0.021135  0.020701   
4  0.098756  0.062228  0.076794  ...  0.073783  0.022212  0.040774  0.043794   

       9994      9995      9996      9997      9998      9999  
0  0.164019  0.088737  0.034574  0.047299  0.065298  0.006800  
1  0.087540  0.047248  0.060795  0.046840  0.028309  0.079752  
2  0.072283  0.011690  0.000000  0.000000  0.088828  0.045069  
3  0.047215  0.028201  0.010773  0.024156  0.046566  0.025094  
4  0.077322  0.038596  0.117033  0.059283  0.054962  0.014998  

[5 rows x 10000 columns]

In [22]:
indices = pd.Series(data.index, index=data['title']).drop_duplicates()

-------------------------------------------------------------------------------------------------------------------------------

## Bagian Ketujuh : Program
Bagian ini adalah bagian utama dari proyek ini, dimana bagian ini adalah bagian yang akan berisi penjelasan dan implementasi dari Wine Matching. Dalam melakukan implementasi, kelompok kami mendapatkan bantuan dari FuzzyWuzzy library. FuzzyWuzzy digunakan untuk string matching dalam program. Keunggulan dari FuzzyWuzzy ini sendiri adalah memudahkan pemyamaan inputan dengan data nantinya. FuzzyWuzzy dapat membantu program mendeteksi kesamaan kata yang memang memiliki arti yang sama walaupun ditulis dalam berbagai macam bentuk.

Contoh: jika kita memasukkan sebuah profesi "Data Scientist" atau "data scientist" atau juga "data science" dan menggunakan data matching yang biasa, mereka dianggap berbeda padahal seharusnya tidak.

FuzzyWuzzy sangat cocok pada data wine, dikarnakan user tidak perlu mengetikkan nama wine sepersis mungkin untuk mendapatkan hasil pencarian. FuzzyWuzzy menerima beberapa parameter antara lain:
1. inputan yang ingin dicari
2. pilihan yang dihasilkan
3. scoring system
4. terkadang limit dari hasil yang diinginkan

In [23]:
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

pilih_satu = ''
def get_pilihan(inputan):
    
    pilihan = data['title']    
    
    # 1. Mencari 5 yang paling mirip dan mempersilahkan user untuk memilih
    mirip = (process.extract(inputan, pilihan, scorer = fuzz.ratio, limit = 5))
    print("\n Berikut adalah 5 wine yang menurut kami mirip, silahkan pilih salah satunya \n")
    print("\n".join(map(str, mirip)))
    
    global pilih_satu
    pilih_satu = input("Masukkan nama wine: ")
    
    print("\n Sepertinya 10 wine ini yang paling mirip", pilih_satu.title(), "\n")
    
    def fuzzy_match(inputan):
    
        pilihan = data['title']    

        # 2. Mengembalikan hasil yang paling mirip teratas
        return process.extractOne(inputan, pilihan, scorer = fuzz.token_sort_ratio, score_cutoff = 50)[0]

    # 3. Assign hasil
    title = fuzzy_match(pilih_satu)

    def get_matching(title, cosine_sim=cosine_sim):

        # 4. Ambil index yang memiliki kemiripan
        idx = indices[title]

        # 5. Pairwise wine yang didapat ke seluruh wine yang ada
        sim_scores = list(enumerate(cosine_sim[idx]))

        # 6. Sorting
        sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

        # 7. Ambil 10 yang paling mirip
        sim_scores = sim_scores[0:11]

        # 8. Ambil index dari wine
        wine_indices = [i[0] for i in sim_scores]

        # Mengembalikan 10 yang paling mirip tersebut
        return data['title'].iloc[wine_indices]

    results = get_matching(title)
    print(results)

In [24]:
pembukaan = input('Wine apa yang anda inginkan ?')
get_pilihan(pembukaan)

Wine apa yang anda inginkan ?black box

 Berikut adalah 5 wine yang menurut kami mirip, silahkan pilih salah satunya 

('Black Box 2013 Malbec (Mendoza)', 46, 1855)
('Black Box 2011 Chardonnay (California)', 39, 2791)
('Vallana 2009  Boca', 37, 7913)
('Brezza 2013  Barolo', 36, 36925)
('Malabaila 2015  Roero', 33, 119138)
Masukkan nama wine: Black Box 2013 Malbec (Mendoza)

 Sepertinya 10 wine ini yang paling mirip Black Box 2013 Malbec (Mendoza) 

20491       Saint-Jean des Sources 2015 Picpoul (Languedoc)
69767     Bowers Harbor 2014 Medium Sweet Riesling (Old ...
6376      Fleur Du Cap 2016 Bergkelder Selection Chenin ...
110338                   Avondale 2013 Cyclus White (Paarl)
81526                      Olivetta 2014 Cortese (Piemonte)
112584    Le Colture 2006 Pianer Extra Dry  (Prosecco di...
42623     Charles Smith 2014 Kung Fu Girl Riesling (Anci...
114949      Rudera 2009 Robusto Chenin Blanc (Stellenbosch)
89749                                  Ottella 2015  Lugana
71671  